Mathlib/Tactic/HashCommandLinter.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+#  `#`-command linter
+
+The `#`-command linter produces a warning when a command starting with `#` is used *and*
+the command emits no message.
+-/
+
+namespace Mathlib.Linter
+
+/--
+The linter emits a warning on any command beginning with `#` that itself emits no message.
+For example, `#guard true` and `#check_tactic True ~> True by skip` trigger a message.
+There is a `whitelist` of silent `#`-command that are allowed.
+-/
+register_option linter.hashCommand : Bool := {
+  defValue := true
+  descr := "enable the `#`-command linter"
+}
+
+namespace HashCommandLinter
+
+open Lean Elab
+
+/-- Gets the value of the `linter.hashCommand` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.hashCommand o
+
+open Command in
+/-- Exactly like `withSetOptionIn`, but recursively discards nested uses of `in`.
+Intended to be used in the `hashCommand` linter, where we want to enter `set_option` `in` commands.
+-/
+private partial def withSetOptionIn' (cmd : CommandElab) : CommandElab := fun stx => do
+  if stx.getKind == ``Lean.Parser.Command.in then
+    if stx[0].getKind == ``Lean.Parser.Command.set_option then
+      let opts ← Elab.elabSetOption stx[0][1] stx[0][2]
+      Command.withScope (fun scope => { scope with opts }) do
+        withSetOptionIn' cmd stx[2]
+    else
+      withSetOptionIn' cmd stx[2]
+  else
+    cmd stx
+
+/-- `whitelist` is the array of `#`-commands that are allowed in 'Mathlib'. -/
+private abbrev whitelist : HashSet String := HashSet.empty
+  |>.insert "#align"
+  |>.insert "#align_import"
+  |>.insert "#noalign"
+
+/-- Checks that no command beginning with `#` is present in 'Mathlib', except for
+* the ones in `whitelist`;
+* the ones that already emit a message.
+
+This means that CI will fail on `#`-commands whether they themselves emit a message or not.
+-/
+def hashCommandLinter : Linter where run := withSetOptionIn' fun stx => do
+--  dbg_trace (← get).messages.msgs.size
+  if getLinterHash (← getOptions) && (← get).messages.msgs.size == 0 then
+--    dbg_trace "linting"
```

Comment:
Did the module check get lost in the mix here, or is this intentional?
```suggestion
  if getLinterHash (← getOptions) && (← get).messages.msgs.size == 0 &&
     (currentModule.getRoot == `Mathlib || currentModule == `test.HashCommandLinter) then
```

Reply:
I intentionally made the linter check everything, including *all* `test` files.  I am not completely convinced by the change, but now you *can* unlint test files, whereas before you did not have the option to lint them, no matter what.

The PR which really makes the linter active (#11028) needed some relatively minor adjustment, but maybe I would leave the discussion of whether `test`s are in scope for the linter to that PR?